File: app/src/main/java/com/mx/gillustrated/adapter/GameListAdapter.java
package com.mx.gillustrated.adapter;

import java.util.List;

import com.mx.gillustrated.R;
import com.mx.gillustrated.vo.GameInfo;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class GameListAdapter extends BaseAdapter {

	private Context mcontext;
	private LayoutInflater layoutInflator;
	private List<GameInfo> list;
	private DespairTouchListener mListener;
	
	public GameListAdapter() {
	}

	public GameListAdapter(Context context, List<GameInfo> items) {
		mcontext = context;
		layoutInflator = LayoutInflater.from(mcontext);
		list = items;
	}
	
	public void setDespairTouchListener( DespairTouchListener listener) {
		mListener = listener;
	}

	@Override
	public int getCount() {
		return list.size();
	}

	@Override
	public Object getItem(int arg0) {
		return list.get(arg0);
	}

	@Override
	public long getItemId(int arg0) {
		return arg0;
	}

	@Override
	public View getView(int arg0, View convertView, ViewGroup arg2) {
		Component component = null;
		
		if (convertView == null) {
			convertView = layoutInflator.inflate(
					R.layout.adapter_game, null);
			component= new Component();
			component.tvName = (EditText) convertView
					.findViewById(R.id.etGameName);
			component.btnSave = (Button) convertView
					.findViewById(R.id.btnDespairModify);
			component.btnDetail = (Button) convertView
					.findViewById(R.id.btnGameDetail);
			convertView.setTag(component);
		}
		else
			component = (Component) convertView.getTag();  
		
		final Component currentComponent = component;
		final int position = arg0;
		try {
			
			component.tvName.setText(list.get(arg0).getName());
			component.btnSave.setOnClickListener( new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					final String name = currentComponent.tvName.getText().toString();
					final int id = list.get(position).getId();
					GameInfo despairInfo = new GameInfo();
					despairInfo.setId(id);
					despairInfo.setName(name);
					despairInfo.setDetail(list.get(position).getDetail());
					mListener.onSaveBtnClickListener(despairInfo);
				}
			});
			component.btnDetail.setOnClickListener( new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					final String name = currentComponent.tvName.getText().toString();
					final int id = list.get(position).getId();
					GameInfo despairInfo = new GameInfo();
					despairInfo.setId(id);
					despairInfo.setName(name);
					mListener.onDetailBtnClickListener(despairInfo);
				}
			});
			
			component.tvName.setOnTouchListener(new View.OnTouchListener() {

				@Override
				public boolean onTouch(View v, MotionEvent event) {
					if (event.getAction() == MotionEvent.ACTION_UP) {
		                 //   index = position;
	                }
	                return false;
				}
	        });
			
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		return convertView;
	}
	
	public interface DespairTouchListener {
		public void onSaveBtnClickListener( GameInfo info );
		public void onDetailBtnClickListener( GameInfo info );
	}
	
	private static class Component{
		 public EditText tvName;  
		 public Button btnSave;
		 public Button btnDetail;  
	}
}

File: app/src/main/java/com/mx/gillustrated/adapter/GameListAdapter.kt
package com.mx.gillustrated.adapter


import com.mx.gillustrated.vo.GameInfo
import android.content.Context
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import com.mx.gillustrated.databinding.AdapterGameBinding

class GameListAdapter constructor(mContext: Context, private val list: List<GameInfo>) : BaseAdapter() {

    private val layoutInflater: LayoutInflater = LayoutInflater.from(mContext)
    private var mListener: DespairTouchListener? = null


    fun setDespairTouchListener(listener: DespairTouchListener) {
        mListener = listener
    }

    override fun getCount(): Int {
        return list.size
    }

    override fun getItem(arg0: Int): Any {
        return list[arg0]
    }

    override fun getItemId(arg0: Int): Long {
        return arg0.toLong()
    }

    override fun getView(arg0: Int, convertViews: View?, arg2: ViewGroup): View {
        var convertView = convertViews
        lateinit var component: AdapterGameBinding

        if (convertView == null) {
            component = AdapterGameBinding.inflate(layoutInflater, arg2, false)
            convertView = component.root
            convertView.tag = component
        } else
            component = convertView.tag as AdapterGameBinding

        try {
            component.etGameName.setText(list[arg0].name)
            component.btnDespairModify.setOnClickListener {
                val name = component.etGameName.text.toString()
                val id = list[arg0].id
                val despairInfo = GameInfo()
                despairInfo.id = id
                despairInfo.name = name
                despairInfo.detail = list[arg0].detail
                mListener!!.onSaveBtnClickListener(despairInfo)
            }
            component.btnGameDetail.setOnClickListener {
                val name = component.etGameName.text.toString()
                val id = list[arg0].id
                val despairInfo = GameInfo()
                despairInfo.id = id
                despairInfo.name = name
                mListener!!.onDetailBtnClickListener(despairInfo)
            }

            component.etGameName.setOnTouchListener { _, event ->
                if (event.action == MotionEvent.ACTION_UP) {
                    //   index = position;
                }
                false
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }

        return convertView
    }

    interface DespairTouchListener {
        fun onSaveBtnClickListener(info: GameInfo)
        fun onDetailBtnClickListener(info: GameInfo)
    }

}
